api/graphite.go

```diff
@@ -0,0 +1,350 @@
+	if err != nil {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, err))
+		return
+	}
+
+	toUnix, err := dur.ParseTSpec(to, now, defaultTo)
+	if err != nil {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, err))
+		return
+	}
+
+	// in MT, both the external and internal api, from is inclusive, to is exclusive
+	// in graphite, from is exclusive and to inclusive
+	// so in this case, adjust for internal api.
+	fromUnix += 1
+	toUnix += 1
+
+	if fromUnix >= toUnix {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, InvalidTimeRangeErr))
+		return
+	}
+	if maxDaysPerReq != 0 && len(targets)*int(toUnix-fromUnix) > maxDaysPerReq*(3600*24) {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, MaxDaysPerReqErr))
+		return
+	}
+
+	reqs := make([]models.Req, 0)
+	parsedTargets := make(map[string]string)
+	patterns := make([]string, 0)
+	type locatedDef struct {
+		def  schema.MetricDefinition
+		node *cluster.Node
+	}
+
+	locatedDefs := make(map[string]map[string]locatedDef)
+	queryForTarget := make(map[string]string)
+	for _, query := range targets {
+		target, consolidateBy, err := parseTarget(query)
+		if err != nil {
+			ctx.Error(http.StatusBadRequest, err.Error())
+			return
+		}
+		parsedTargets[target] = consolidateBy
+		patterns = append(patterns, target)
+		queryForTarget[target] = query
+		locatedDefs[target] = make(map[string]locatedDef)
+
+		// metricDefs only get updated periodically, so we add a 1day (86400seconds) buffer when
+		// filtering by our From timestamp.  This should be moved to a configuration option,
+		// but that will require significant refactoring to expose the updateInterval used
+		// in the MetricIdx.
+		seenAfter := int64(fromUnix)
+		if seenAfter != 0 {
+			seenAfter -= 86400
+		}
+		nodes, err := s.MetricIndex.Find(ctx.OrgId, target, seenAfter)
+		if err != nil {
+			rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, err))
+			return
+		}
+		for _, node := range nodes {
+			for _, def := range node.Defs {
+				locatedDefs[target][def.Id] = locatedDef{def, cluster.ThisNode}
+			}
+		}
+	}
+
+	for target, ldefs := range locatedDefs {
+		for _, locdef := range ldefs {
+			def := locdef.def
+			consolidator, err := consolidation.GetConsolidator(&def, parsedTargets[target])
+			if err != nil {
+				rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, err))
+				return
+			}
+			// query is like foo.bar or foo.* or consolidateBy(foo.*,'sum')
+			// target is like foo.bar or foo.*
+			// def.Name is like foo.concretebar
+			// so we want query to contain the concrete graphite name, potentially wrapped with consolidateBy().
+
+			query := strings.Replace(queryForTarget[target], target, def.Name, -1)
+			reqs = append(reqs, models.NewReq(def.Id, query, locdef.node, fromUnix, toUnix, maxDataPoints, uint32(def.Interval), consolidator))
+		}
+	}
+	if len(reqs) == 0 {
```

Comment:
why would this ever happen? didn't we already validate that there must be at least one non-empty target in the request, so there should be a req ?